### PR TITLE
KDTree Data Structure for Center and Corner Nodes

### LINKED ISSUE #496: Implement `KdTree` built from Cartesian x/y/z Coordinates
To support regridding using cartesiain coordinates, we should implement a `KdTree` with the same API as our `BallTree`

### PR DESCRIPTION
**Pull request by aaronzedwick**
Closes #496

## Overview

Adds a KDTree data structure for center nodes (`Mesh2_face_cart_x`, `Mesh2_face_cart_y`, `Mesh2_face_cart_z`) and corner nodes (`Mesh2_node_cart_x`, `Mesh2_node_cart_y`, `Mesh2_node_cart_z`)  for grid files. This can be used for nearest neighbor queries as well as querying in a specific radius. This implementation mirrors the BallTree data structure #406 and uses `sklearn.neighbors.KDTree` with a minkowski distance for querying.  

## Expected Usage

```Python
import uxarray as ux

grid_path = "/path/to/grid.nc"

uxgrid = ux.open_grid(grid_path)

# find the nearest neighbor to (0.0, 0.0, 1.0) using corner nodes
d, ind = uxgrid.get_kd_tree(tree_type="nodes").query([0.0, 0.0, 1.0])

# find the 5 nearest neighbors to (0.0, 0.0, 1.0) using face centers
d, ind = uxgrid.get_kd_tree(tree_type="face centers").query([1.0, 0.0, 0.0], k=5)

# find all face centers within a radius of 80.00 from (0.0, 0.0, 1.0)
d, ind = uxgrid.get_kd_tree(tree_type="face centers").query_radius([0.0, 0.0, 1.0], r=80.00)
```

## PR Checklist

**General**
- [x] An issue is linked created and linked
- [x] Add appropriate labels
- [x] Filled out Overview and Expected Usage (if applicable) sections

**Testing**
- [x] Adequate tests are created if there is new functionality
- [x] Tests cover all possible logical paths in your function
- [x] Tests are not too basic (such as simply calling a function and nothing else)

**Documentation**
- [x] Docstrings have been added to all new functions
- [x] Docstrings have updated with any function changes
- [x] Internal functions have a preceding underscore (`_`) and have been added to `docs/internal_api/index.rst`
- [x] User functions have been added to `docs/user_api/index.rst`

**Examples**
- [x] Any new notebook examples added to `docs/examples/` folder
- [x] Clear the output of all cells before committing
- [x] New notebook files added to `docs/examples.rst` toctree
- [x] New notebook files added to new entry in `docs/gallery.yml` with appropriate thumbnail photo in `docs/_static/thumbnails/`

### CODE REVIEW

**Review by rajeeja — CHANGES_REQUESTED** [review]

File: uxarray/grid/neighbors.py [hunk 1]
```diff
@@ -268,3 +493,25 @@ def _prepare_xy_for_query(xy, use_radians):
         xy = np.deg2rad(xy)
 
     return xy
+
+
+def _prepare_cart_xyz_for_query(xyz):
```
Comment (philipc2) [inline uxarray/grid/neighbors.py:498]:
```suggestion
def _prepare_xyz_for_query(xyz):
```

this name should suffice

File: uxarray/grid/neighbors.py [hunk 2]
```diff
@@ -3,11 +3,236 @@
...
+            Distance array that keeps the distances of all neighbors within some radius to the entries from xyz in each row
+        ind : ndarray of shape (xyz.shape[0], k), dtype=INT_DTYPE
+            Index array that keeps the indices of all neighbors within some radius to the entries from xyz in each row
+        """
+
+        if r < 0.0:
+            raise AssertionError(
+                f"The value of r must be greater than or equal to zero.")
+
+        r = np.deg2rad(r)
```
Comment (philipc2) [inline uxarray/grid/neighbors.py:190]: I'd be careful with this.

The reason we did the `deg2rad` conversion is because our `BallTree` used the haversine distance, which works in radians.

Here we are using the mikowski distance in Cartesian space, I'm not sure what the physical meaning of our "r" is. Can you look into it?

**Review by philipc2 — CHANGES_REQUESTED** [review]

File: uxarray/grid/neighbors.py (hunk 2 above)
Comment (aaronzedwick) [reply uxarray/grid/neighbors.py:190]: Yeah, I think you are right, it does make sense that r should be in whatever measurement you have for your input grid.

File: uxarray/grid/neighbors.py [hunk 3]
```diff
@@ -3,11 +3,236 @@
...
+                     count_only: Optional[bool] = False,
+                     sort_results: Optional[bool] = False):
+        """Queries the tree for all neighbors within a radius ``r``.
+
+        Parameters
+        ----------
+        xyz : array_like
+           coordinate pairs in cartesian (x, y, z) to query
+        r: distance in degrees within which neighbors are returned
+            r can be a single value , or an array of values of shape x.shape[:-1] if different radii are desired for each point.
```
[the remark this answers is not in the document]
Comment (aaronzedwick) [reply uxarray/grid/neighbors.py:170]: @philipc2 I haven't actually seen the documentation for having multiple radii, are we sure this is something that works? I sent in an array of points and a matching array of radii but it didn't accept that input.

[the remark this answers is not in the document]
Comment (philipc2) [reply uxarray/grid/neighbors.py:170]: Good catch, I think I may have copied this docstring section for `sklearn` directly without thinking about it. I'm not sure if it works. I think it would be OK to remove that part from both the KD and Ball Tree and we can investigate it in the future.

**Comment by aaronzedwick** [comment]
@philipc2 I think I updated the user api incorrectly causing the docs to fail, anything  you see that would be causing the problem? I just added what you did from the balltree, but using the kdtree instead.

File: docs/user_api/index.rst [hunk 4]
```diff
@@ -175,6 +176,10 @@ Attributes
    Grid.Mesh2_edge_nodes
    Grid.Mesh2_face_edges
    Grid.antimeridian_face_indices
+   Grid.corner_node_balltree
+   Grid.center_node_balltree
+   Grid.corner_node_kdtree
+   Grid.center_node_kdtree
```
Comment (philipc2) [inline docs/user_api/index.rst:182]: @aaronzedwick 

I think these are causing your issue, we don't actually have these attributes.

File: docs/user_api/index.rst [hunk 5]
```diff
@@ -217,13 +222,22 @@ BallTree Data Structure
 
    grid.BallTree
 
+KDTree Data Structure
+-----------------------
```
Comment (philipc2) [inline docs/user_api/index.rst:226]:
```suggestion
KDTree Data Structure
------------------------
```

**Review by philipc2 — APPROVED** [review]
Nice work! Please run the pre-commit though!
